cpacker/MemGPT
# feat: Support pydantic models as parameters to custom functions

**Please describe the purpose of this pull request.**
Add support for pydantic models (also nested ones) for custom functions

**How to test**
Custom function for testing:
```python
from typing import List, Any

from pydantic import Field, BaseModel


class RowData(BaseModel):
    row: List[Any] = Field(..., description="The values for each row")
    citation: str = Field(
        ..., description="The citation for this row from the original source data"
    )


class Dataframe(BaseModel):
    """
    Class representing a dataframe. This class is used to convert
    data into a frame that can be used by pandas.
    """

    name: str = Field(..., description="The name of the dataframe")
    data: List[RowData] = Field(
        ...,
        description="Correct rows of data aligned to column names, Nones are allowed",
    )
    columns: List[str] = Field(
        ...,
        description="Column names relevant from source data, should be in snake_case",
    )

    def to_pandas(self):
        import pandas as pd

        columns = self.columns + ["citation"]
        data = [row.row + [row.citation] for row in self.data]

        return pd.DataFrame(data=data, columns=columns)


class Database(BaseModel):
    """
    A set of correct named and defined tables as dataframes
    """

    tables: List[Dataframe] = Field(
        ...,
        description="List of tables in the database",
    )


def save_database(self, database: Database):
    """
    Save the database and return string representation.

    Args:
        database: a database object
    """
    dataframes = ""
    for df in database.tables:
        dataframes += df.name + "\n"
        dataframes += str(df.to_pandas()) + "\n"
    return dataframes
```



**Have you tested this PR?**
Made some tests to see if it works.

![Screenshot 2024-01-17 223932](https://github.com/cpacker/MemGPT/assets/24946356/3f06b2ef-4ef0-4238-b63b-ff79d21e5310)
![Screenshot 2024-01-17 223957](https://github.com/cpacker/MemGPT/assets/24946356/7ae4503c-cd11-464e-a98a-cf2a43f1460e)
